drivers/i2c/i2c_bitbang.c

```diff
@@ -264,6 +264,46 @@ int i2c_bitbang_transfer(struct i2c_bitbang *context,
 	return result;
 }
 
+int i2c_bitbang_recover_bus(struct i2c_bitbang *context)
+{
+	int i;
+
+	/*
+	 * The I2C-bus specification and user manual (NXP UM10204
+	 * rev. 6, section 3.1.16) suggests the master emit 9 SCL
+	 * clock pulses to recover the bus.
+	 *
+	 * The Linux kernel I2C bitbang recovery functionality issues
+	 * a START condition followed by 9 STOP conditions.
+	 *
+	 * Other I2C slave devices (e.g. Microchip ATSHA204a) suggest
+	 * issuing a START condition followed by 9 SCL clock pulses
+	 * with SDA held high/floating, a REPEATED START condition,
+	 * and a STOP condition.
+	 *
+	 * The latter is what is implemented here.
```

Comment:
I am not convinced regarding using such procedure for the general case. The goal of the I2C bus recovery procedure should be only to get back the control over the SDA line, so that the master is able to generate the STOP condition (what is impossible while a slave holds the SDA line low) and this way to finish the transaction that apparently was interrupted earlier (if there is some other cause of the SDA line being low, there is probably nothing that firmware can do to help in this regard). Thus, the SCL pulses should be generated only until the SDA line gets high. The whole thing is quite nicely described in the application note [AN-686](https://www.analog.com/media/en/technical-documentation/application-notes/54305147357414AN686_0.pdf) from Analog Devices (titled "Implementing an I2C® Reset").

When it comes to Nordic drivers, we've got such function already [implemented](https://github.com/zephyrproject-rtos/hal_nordic/blob/master/nrfx/drivers/src/nrfx_twi_twim.c#L42) (and tested on real life cases that were motivation for creating it), so it rather won't be suitable for us to use the i2c_bitbang library instead.